### PR TITLE
feat(backend): new server API to read run log. Fixes #4468

### LINKED ISSUE #4468: Rest API for streaming pipeline logs
Having a Rest API for log can be useful is we are using the CLI or SDK client. Is there any plan for adding a REST API to stream logs or it's by design due to security or performance concern? Thanks.

### PR DESCRIPTION
**Pull request by rafalbigaj**
**Description of your changes:**

Resolves issue: #4468

The new backend API to read run log is proposed in place of the existing BFF endpoint.

New endpoint: GET /apis/v1beta1/runs/{run_id}/nodes/{node_id}/log

Having a Rest API for log can be useful is we are using the CLI or SDK client. 
Additionally, in some cases, the server logic responsible for log management is better placed in the API Server.
The new UI Server option `STREAM_LOGS_FROM_SERVER_API` enables streaming from API Server.

**Checklist:**
- [X] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?

    [Learn more about cherry-picking updates into the release branch](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#cherry-picking-pull-requests-to-release-branch).

### CODE REVIEW

**Comment by rafalbigaj** [comment]
Original PR: https://github.com/kubeflow/kfp-tekton/pull/299

**Comment by rafalbigaj** [comment]
@rmgogogo @Tomcli thanks for your support!
I additionally delivered changes in tests.

**Comment by rafalbigaj** [comment]
@rmgogogo after delivering the test fixes lgtm label was removed.
Would you be able to mark it again? Thanks!

**Comment by Bobgy** [comment]
/cc @eterna2 
for review, who contributed to the log endpoint in KFP UI Server

File: frontend/server/app.ts
Comment (Bobgy) [inline frontend/server/app.ts:153]: nit: comment argo nodeId is just podname, maybe it can change in the future

File: backend/src/apiserver/main.go
Comment (Bobgy) [inline backend/src/apiserver/main.go:149]: Please use the existing convention of writing *.proto definitions for the endpoint and generate clients/server stubs for the endpoint. The benefits:
* KFP API SDK is auto generated
* KFP API Reference is auto generated

Feel free to let me know if you cannot find related resources to use the existing pattern

**Review by Bobgy — COMMENTED** [review]
Hi @rafalbigaj, thank you for your contribution!
I'd suggest proceed like the following:
* discuss and confirm on user facing features this PR provides
* discuss and confirm on high level tech choices (e.g. use proto)
* detailed review
* merge

**Comment by Bobgy** [comment]
For "user facing features this PR provides":

Why did you do log parsing (e.g. k8s logs, json format logs...) in API server? What's your use-case? Can it be done in other services? Because it doesn't look like KFP's main responsibility.

For tech choices:
My major concern is https://github.com/kubeflow/pipelines/pull/4493#discussion_r491859229, is the problem this endpoint returns blob data?

File: backend/src/apiserver/main.go
Comment (rafalbigaj) [reply backend/src/apiserver/main.go:149]: @Bobgy Thanks for your comment and suggestions.
Before I switch to gRPC and http proxy I'd like to ask about support for realtime log streaming from k8s POD, which is quite important for us.

1. At first k8s logs API is based on HTTP chunked transfer encoding and the same is proposed here. k8s logs are by nature unstructured (only timestamp can optionally be extracted).
2. Secondly, the priority for me was the raw http endpoint that can be easily used (even with `curl`) to download logs in the original form.

My understanding is that any solution based on proto would add complexity to the message structure and downloaded log would require post-processing.

I would love to see a good and efficient pattern that addresses all requirements.
Thanks!

**Comment by rafalbigaj** [comment]
@Bobgy this endpoint is expected to work in quite diversified solutions:
1. Argo with build-in log archival mechanism.
2. Tekton with [custom logging to s3](https://github.com/kubeflow/kfp-tekton/tree/master/samples/logging_s3) or this [multi-s3 fluent-bit plugin](https://github.com/rafalbigaj/fluent-bit-multi-s3). In case of tekton there is no native support for the log archive.

The second option simply copies POD logs to s3 bucket as they are in the json, cri-o or raw format (without any post-processing, the format depends on the container engine). That creates a need to parse them and extract timestamp and log message in API Server.

If you find this logic redundant for KFP API, it needs to be moved to the code that persists logs on s3 in the raw format. The timestamp information would be lost, until we agree on a common format for both cases (argo and tekton with custom solutions).

File: frontend/server/app.ts
Comment (eterna2) [inline frontend/server/app.ts:162]: If this API is implemented in kfp api server, I think we can deprecate this part as  it is essentially identical to the golang version except it doesn't handle the log formats (it just stream outright - but not difficult to add a transform to the stream thou).

Comment (rafalbigaj) [reply frontend/server/app.ts:162]: @eterna2 that's correct. We can deprecate, but we need to remember that streaming through server API means one hop more (UI -> API proxy -> Server API -> k8s API). Even though, I don't see significant performance impact.

File: backend/src/apiserver/main.go
Comment (Bobgy) [reply backend/src/apiserver/main.go:149]: These are great points! Totally understood your decision now.
Sounds good to me.

File: frontend/server/app.ts
Comment (Bobgy) [reply frontend/server/app.ts:162]: I don't see a big concern with one more hop too, a lot of requests have this extra hop, but they are also working well.

**Comment by Bobgy** [comment]
@rafalbigaj Sorry, I didn't have much experience with different container runtimes before. Are these formats "the json, cri-o or raw format" emitted as log format because of the container runtime, like containerd etc? or are they a choice of user code in the container?

If they are sort of standard, do we have existing go libraries that can process them?

Another question: I am not seeing a kubernetes client based log streaming implementation (like kubectl logs xxx), do you have a plan for it? I think it is required to completely replace the UI logs api implementation.

File: frontend/server/app.ts
Comment (rafalbigaj) [reply frontend/server/app.ts:162]: Agreed. Let's do it in the subsequent PR once this is merged.

**Comment by rafalbigaj** [comment]
@Bobgy Indeed, the log format depends on the container runtime and I haven't found any go tool that could be used here. I based this implementation mostly on fluent-bit configurations (regexp parsers).

I've added support for `follow` query parameter in `GET /apis/v1beta1/runs/{run_id}/nodes/{node_id}/log`, even though I don't see usage from UI side. For SDK/CLI it would be nice to have.

**Comment by eterna2** [comment]
I forgot to mention something. This is something I always wanted to do when I got time, but never gone around doing it.

The workflow CR is removed after a sometime by the persistent agent. So theoretically we should not be trying to get the node status from the CR. The workflow CR is persisted inside the MySQL run table. So we should be getting the artifact location from the run table instead.

However, I think the run table is indexed by the run id, which is different from the workflow id. So I think this is something we need to look at in the future, for a more proper log management strategy.

**Comment by Bobgy** [comment]
@rafalbigaj Sorry, I've just been back from vacation.
Can I confirm if this is just for log archive API? do you intend to support kubernetes logs API, so it could be used for real time tasks.

**Comment by rafalbigaj** [comment]
@Bobgy it supports both: runtime logs from kubernetes API as well as the archived logs from object storage.

**Comment by rafalbigaj** [comment]
@Bobgy could we proceed with this PR?

File: backend/src/apiserver/server/run_log_server.go
Comment (Bobgy) [inline backend/src/apiserver/server/run_log_server.go:1]: nit: `2020`
Also, we are open to marking these files as `Copyright 2020 The Kubeflow Authors` if you'd prefer that.

File: backend/src/apiserver/main.go
Comment (Bobgy) [inline backend/src/apiserver/main.go:149]: Shall we mark the api as `v1alpha1` initially?

File: backend/src/apiserver/config/config.json
Comment (Bobgy) [inline backend/src/apiserver/config/config.json:17]: We are moving towards env var friendly config names, so I'd suggest the following:
```
  "ARCHIVE_CONFIG": {
    "LOG_FILE_NAME": "main.log",
    "LOG_PATH_PREFIX": "/artifacts"
  },
```

and when overriding the config from env var, we can use `ARCHIVE_CONFIG_LOG_FILE_NAME`

File: backend/src/apiserver/client_manager.go
Comment (Bobgy) [inline backend/src/apiserver/client_manager.go:383]: common.GetStringConfigWithDefault is already getting from env var. You can read viper's documentation.

For `ArchiveConfig.LogFileName`, corresponding env var name will be `ARCHIVECONFIG_LOGFILENAME`, so I'd suggest looking at my other comment to make it env var friendly.

File: backend/src/apiserver/archive/log.go
Comment (Bobgy) [inline backend/src/apiserver/archive/log.go:56]: I think you mentioned some of these lines of code were "copied" from a library with similar functionality. Can you include its full license file around those lines?

**Review by Bobgy — COMMENTED** [review]
/lgtm

The PR moves core feature like fetching logs to KFP API Server, that's a great contribution!
I especially appreciate your efforts to build feature gates that default to not use this new API.

My major comment is marking this API as alpha initially, when people deploy and verify its features by switching over to it. We can move forward its stability over time and eventually it should replace the api on ui server.

For next steps, I'd probably suggest hand-writing a swagger.json definition for your api, so that we can auto-generate a client for it. You can find existing auto-generated swagger json definitions in https://github.com/kubeflow/pipelines/tree/master/backend/api/swagger.

**Comment by Ark-kun** [comment]
It would be great if this API would properly return the logs for the nodes reused from cache.
For the reused Argo nodes, the logs are in the same main-logs artifact, so it just need to be used.

**Comment by Bobgy** [comment]
> It would be great if this API would properly return the logs for the nodes reused from cache.
> For the reused Argo nodes, the logs are in the same main-logs artifact, so it just need to be used.

Agreed, this is a chance to improve caching.
Can you inform us how to find a cached pod's previous pod name and workflow name?

I think that can be a follow-up PR, what do you think @Ark-kun ?

File: backend/src/apiserver/main.go
Comment (rafalbigaj) [reply backend/src/apiserver/main.go]: @Bobgy Sure, I changed the path

File: backend/src/apiserver/config/config.json
Comment (rafalbigaj) [reply backend/src/apiserver/config/config.json]: We have env var friendly names now

File: backend/src/apiserver/client_manager.go
Comment (rafalbigaj) [reply backend/src/apiserver/client_manager.go]: Done, thanks for comment.

File: backend/src/apiserver/archive/log.go
Comment (rafalbigaj) [reply backend/src/apiserver/archive/log.go:59]: I mentioned about fluent-bit parsers and referenced Apache 2.0 licence

**Comment by rafalbigaj** [comment]
@Bobgy Thank you very much for your comments.
I would love to continue the work on this new capability in the separate PR.

File: backend/src/apiserver/archive/log.go
Comment (Bobgy) [inline backend/src/apiserver/archive/log.go:57]: Would you mind copying their full Apache License inline?

**Review by Bobgy — APPROVED** [review]
Thanks for the update! There's one last comment.
We've waited enough time for others to review, so I think we can proceed once the last comment is resolved.
/approve
/lgtm
/hold

**Comment by Bobgy** [comment]
Also, can you please resolve the merge conflict?

File: backend/src/apiserver/archive/log.go
Comment (rafalbigaj) [reply backend/src/apiserver/archive/log.go]: Sure, license inlined

**Comment by rafalbigaj** [comment]
@Bobgy merged conflicts resolved. We should be good.

**Review by Bobgy — APPROVED** [review]
/unhold
/lgtm
Thank you @rafalbigaj !

**Comment by eshaingle** [comment]
Hello

I am new to Kubeflow pipelines. May be my question a basic one but just want to know what will be the node id here?
Also, is this new endpoint working with 1.3 version?

**Pull request merged**
